src/runtime/vm/executable.cc

```diff
@@ -0,0 +1,726 @@
+    oss << "# Parameters: "<< std::endl;
+    for (const auto& param : func.params) {
+      oss << param << " ";
+    }
+    oss << std::endl;
+
+    // Print the instructions of a `VMFunction`.
+    // The part after ";" is the instruction in text format.
+    oss << "hash, opcode, fields # inst(text):"<< std::endl;
+    for (const auto& instr : func.instructions) {
+      const auto& serialized_instr = SerializeInstruction(instr);
+      oss << std::hex << "0x" << serialized_instr.Hash() << " "
+          << std::dec << serialized_instr.opcode << " ";
+      for (auto it : serialized_instr.fields) {
+        oss << it << " ";
+      }
+      oss << "  # " << instr;
+      if (oss.str().back() != '\n') oss << std::endl;
+    }
+  }
+
+  return oss.str();
+}
+
+std::string Executable::Stats() const {
+  std::ostringstream oss;
+  oss << "Relay VM executable statistics:" << std::endl;
+
+  // Get the number of constants and the shape of each of them.
+  oss << "  Constant shapes (# " << constants.size() << "): [";
+  for (const auto& it : constants) {
+    const auto* cell = it.as<TensorObj>();
+    CHECK(cell);
+    runtime::NDArray data = cell->data;
+    const auto& shape = data.Shape();
+
+    // Scalar
+    if (shape.empty()) {
+      oss << "scalar, ";
+      continue;
+    }
+
+    oss << "[";
+    for (auto s : shape) {
+      oss << s << ", ";
+    }
+    oss.seekp(-2, oss.cur);
+    oss << "], " << std::endl;
+  }
+  if (!constants.empty()) oss.seekp(-2, oss.cur);
+  oss << "]" << std::endl;
+
+  // Get the number of globals and the name of each of them.
+  oss << "  Globals (#" << global_map.size() << "): [";
+  for (const auto& it : global_map) {
+    oss << "(\"" << it.first << "\", " << it.second << ")" << ", ";
+  }
+  if (!global_map.empty()) oss.seekp(-2, oss.cur);
+  oss << "]" << std::endl;
+
+  // Get the number of primitive ops and the name of each of them.
+  oss << "  Primitive ops (#" << primitive_map.size() << "): [";
+  std::vector<std::string> prim_ops;
+  for (const auto& it : primitive_map) {
+    auto packed_index = static_cast<size_t>(it.second);
+    if (prim_ops.size() <= packed_index) {
+      prim_ops.resize(packed_index + 1);
+    }
+    prim_ops[packed_index] = it.first;
+  }
+  for (const auto& it : prim_ops) {
+    oss << it << ", ";
+  }
+  if (!prim_ops.empty()) oss.seekp(-2, oss.cur);
+  oss << "]" << std::endl;
+
+  return oss.str();
+}
+
+TVMByteArray Executable::Save() {
+  // Initialize the stream object.
+  strm_ = new dmlc::MemoryStringStream(&code_);
+
+  uint64_t header = kTVMVMBytecodeMagic;
+  strm_->Write(header);
+  std::string version = TVM_VERSION;
+  strm_->Write(version);
+
+  // Global section.
+  SaveGlobalSection();
+
+  // Constant section.
+  SaveConstantSection();
+
+  // Primitive names.
+  SavePrimitiveOpNames();
+
+  // Code section.
+  SaveCodeSection();
+
+  TVMByteArray arr;
+  arr.data = code_.c_str();
+  arr.size = code_.length();
+  return arr;
```

Comment:
delete strm_ before return or in the deconstructor to prevent memory leak. Or just use a local var and pass the ref to helper function. I prefer the latter one. Same for the `Load` function.